src/Cassandra/TabletMap.cs

```diff
@@ -0,0 +1,238 @@
+                    LastToken == other.LastToken &&
+                    EqualityComparer<List<HostShardPair>>.Default.Equals(Replicas, other.Replicas);
+        }
+
+        public override int GetHashCode()
+        {
+            // Manual hash code implementation for .NET Standard 2.0
+            int hash = 17;
+            hash = hash * 23 + (KeyspaceName != null ? KeyspaceName.GetHashCode() : 0);
+            hash = hash * 23 + (TableId != null ? TableId.GetHashCode() : 0);
+            hash = hash * 23 + (TableName != null ? TableName.GetHashCode() : 0);
+            hash = hash * 23 + FirstToken.GetHashCode();
+            hash = hash * 23 + LastToken.GetHashCode();
+            hash = hash * 23 + (Replicas != null ? Replicas.GetHashCode() : 0);
+            return hash;
+        }
+
+        public override string ToString()
+        {
+            return $"Tablet{{keyspaceName='{KeyspaceName}', tableId={TableId}, tableName='{TableName}', firstToken={FirstToken}, lastToken={LastToken}, replicas={Replicas}}}";
+        }
+    }
+
+    public class TabletMap
+    {
+        private static readonly Logger Logger = new Logger(typeof(TabletMap));
+        private static readonly IReadOnlyList<HostDummy> EMPTY_LIST = new List<HostDummy>();
+
+        private readonly ConcurrentDictionary<KeyspaceTableNamePair, TabletSet> mapping;
+        private readonly ClusterManager cluster;
+
+        public TabletMap(ClusterManager cluster, ConcurrentDictionary<KeyspaceTableNamePair, TabletSet> mapping)
+        {
+            this.cluster = cluster;
+            this.mapping = mapping;
+        }
+
+        public static TabletMap EmptyMap(ClusterManager cluster)
+        {
+            return new TabletMap(cluster, new ConcurrentDictionary<KeyspaceTableNamePair, TabletSet>());
+        }
+
+        public IDictionary<KeyspaceTableNamePair, TabletSet> GetMapping() => mapping;
+
+        public IReadOnlyList<HostDummy> GetReplicas(string keyspace, string table, long token)
+        {
+            var key = new KeyspaceTableNamePair(keyspace, table);
+
+            if (!mapping.TryGetValue(key, out var tabletSet))
+            {
+                Logger.Info("No tablets for {keyspace}.{table} in mapping.", keyspace, table);
+                return EMPTY_LIST;
+            }
+
+            tabletSet.Lock.EnterReadLock();
+            try
+            {
+                var row = tabletSet.Tablets.FirstOrDefault(t => t.LastToken >= token);
+                if (row == null || row.FirstToken >= token)
+                {
+                    Logger.Info("Could not find tablet for {keyspace}.{table} owning token {token}.", keyspace, table, token);
+                    return EMPTY_LIST;
+                }
+
+                var replicas = new List<HostDummy>();
+                foreach (var hostShardPair in row.Replicas)
+                {
+                    var replica = cluster.Metadata.GetHostDummy(hostShardPair.HostDummy);
```

Comment:
let's try to subscribe to `cluster.HostRemoved` and `cluster.HostAdded`, keep list of the hosts instead of hostids